### PR TITLE
Add not entirely complete definition of Item struct and related enums.

### PR DESCRIPTION
**Pull request by Predelnik**
### Naming:

1. I used `item_` prefix for all item related enums, it actually has 2 benefits: 1) it's easier to find and group them 2) fields in `Item` stuct could be named the same way as enums after removing prefix. I think it looks a bit weird only for a field `id`.  However there's a slight problem with naming `unique_item_id` field which I for now named rather clumsily `unique_id`. 
Speaking in general such problems in the future could be alleviated by adding to all enum names suffix like `_t` or adding to all struct fields prefix like `m_` but I'm not sure if decisions like that are really good.

2. I used `category` rather than `group` since it seems a bit more appropriate for such huge subset of items.

3. I added `inv_` to graphics id to separate inventory graphics from drop graphics, and used abbreviation to not make enum names too long.

4. I kept name of `speical_effect` enum unchanged since I quite surprisingly gave it the same name. I've done some changes to naming ,in particular by adding to faster attack and hit recovery its degree as a number because terms `fast`, `faster` are bit imprecise for me but I'm not super sure, maybe we should stick to in-game terms for these ones.

5. `Item_code` I had no idea how to name it so I used the one used in freeablo, I think name `code` is as good as any.

6. I renamed `ability_type` to `effect_type`, since I believe abilities usually have a bit different meaning in rpg and it also aligns nicely with `special_effect`. In my own notes I named it `affix_effect_type` but I think that since it also describes unique item ability, simply `effect` is probably better.
So I took a bit of freedom in naming enum items here, for all simple +/- effect I tried to use direct suffixes `PLUS`, `MINUS` instead of something like `increase` or `enhanced` which in my opinion makes it a bit simpler to read. Also it's clear to see that difference between some adjacent effects is only in sign.

7. About Item struct itself. 

* As I mentioned I tried to avoid using `item_` in field names for the most part since it's sort of redundant.

* I personally like the word `bonus` since I believe it tells that it adds something, and that sum from all items is calculated eventually. Perhaps negative bonus is not really common usage but it's still understandable.

* For `charges` and `spell_id` I explicitly added that it's related to staff

* For things which +% of something I like to add `percent` explicitly to name, to understand what they do easier.

* I explicitly named basic item name and price `unindetified` which I believe clearer tells the purpose of why they are kept inside `Item` struct.

### Some TODOs

1. I don't quite know myself what the following fields in `Item` mean:

* `drop_anim_update`
* `inactive`
* `drop_state` and its enum
* `draw_quest_item`

And I do believe that some of them require comment at least. Actually I'm fine with leaving them as it is and later perhaps me or someone else will add comment or rename them if will find it appropriate.

2. Perhaps some other fields require comment too. Some enums may benefit from comments also, like for special effects or effect types adding item which has them as a point of reference, if they are rare. But possibly it could be added in the future.

3. Possibly I should remove part in comment about `Q26.6` from health, mana bonus because it looks too obscure.

4.  I've tried adding `\\ unused`  to effects missing in game but probably something longer like `\\ missing in game` should be used instead for more clarity.

5. Should we really call `reserved` fields which are most likely padding? Should perhaps bit-fields be used for adding `24` - bit padding. (like `int reserved : 24`)?

6. Some enums clearly has underlying type that just 8-bit but some look like 32-bit (or were they just stored in int). It's important to denote that somehow, otherwise `struct` layout will be incomplete

7. Should I also used types like `bool32_t` for types which denote boolean values? I think that adding prefix `is_` sort of makes clear that they represent bool.

### CODE REVIEW

**Comment by mewmew** [comment]
Wow! This is a first-class pull request. I love that you've documented the thought processes that went into naming, and how you tried to make things consistent and resolve ambiguities.

I'll add specific comments for each bullet item below.

**Comment by mewmew** [comment]
# Naming

## 1.

> I used item_ prefix for all item related enums, it actually has 2 benefits: 1) it's easier to find and group them 2) fields in Item stuct could be named the same way as enums after removing prefix. I think it looks a bit weird only for a field id. However there's a slight problem with naming unique_item_id field which I for now named rather clumsily unique_id.
Speaking in general such problems in the future could be alleviated by adding to all enum names suffix like _t or adding to all struct fields prefix like m_ but I'm not sure if decisions like that are really good.

I agree that an `item_` prefix for the enums may be a good thing.

Using `id` and `unique_id` is a good way to resolve the ambiguity. I think we may have to revisit the general idea of dropping the prefix for struct member names when working with larger structs than Item. For instance, I believe the `Player` struct (21720 bytes in size) may contain several xxx_id enums. However, lets use this convention for now, and change it when we run into larger ambiguities.

I do not wish to use `_t` suffixes or `m_` prefixes. Feels to C++ influenced.

## 2.

> I used category rather than group since it seems a bit more appropriate for such huge subset of items.

Agreed, category is better than group for this purpose.

## 3.

> I added inv_ to graphics id to separate inventory graphics from drop graphics, and used abbreviation to not make enum names too long.

An inv_ prefix and a drop_ prefix is good to differentiate between the two. My only concern is that inv_ graphics are also used in the belt, and for the cursor, besides the inventory.

## 4.

> I kept name of speical_effect enum unchanged since I quite surprisingly gave it the same name. I've done some changes to naming ,in particular by adding to faster attack and hit recovery its degree as a number because terms fast, faster are bit imprecise for me but I'm not super sure, maybe we should stick to in-game terms for these ones.

Haha, good you came up with the same name.

I would say, *great minds think alike*.

My brother would counter with, *small minds rarely differ*.

Regarding `fast`, `faster`, and numeric suffixes. In this case, I feel the `fast`, `faster` actually is more precise, since those are the in game descriptions used. An incremental numeric ID loses this. I'd still be happy to add `// +3% attack speed` or something along those lines as comments.

## 5.

> Item_code I had no idea how to name it so I used the one used in freeablo, I think name code is as good as any.

I know, we were also struggling with this one. Lets use code for now, and perhaps profound insight will strike us at a later point.

## 6.

> I renamed ability_type to effect_type, since I believe abilities usually have a bit different meaning in rpg and it also aligns nicely with special_effect. In my own notes I named it affix_effect_type but I think that since it also describes unique item ability, simply effect is probably better.
So I took a bit of freedom in naming enum items here, for all simple +/- effect I tried to use direct suffixes PLUS, MINUS instead of something like increase or enhanced which in my opinion makes it a bit simpler to read. Also it's clear to see that difference between some adjacent effects is only in sign.

I agree, effect_type is better and brings the concept closer to special_effect. Ability is more along the lines of magic in RPGs.

Sure, PLUS and MINUS suffixes are fine by me.

## 7.

> About Item struct itself.

> As I mentioned I tried to avoid using item_ in field names for the most part since it's sort of redundant.

Noted.

> I personally like the word bonus since I believe it tells that it adds something, and that sum from all items is calculated eventually. Perhaps negative bonus is not really common usage but it's still understandable.

Yes, this was actually one of the things I liked when reading your initial Item notes in txt format. Lets go for the _bonus suffix. Negative bonuses are fine, and I love that Diablo makes good use of them with items that give you both strengths in one realm and make you weaker in others.

> For charges and spell_id I explicitly added that it's related to staff

Sure, sounds good.

> For things which +% of something I like to add percent explicitly to name, to understand what they do easier.

Definitely, a percent suffix makes sense in these cases.

> I explicitly named basic item name and price unindetified which I believe clearer tells the purpose of why they are kept inside Item struct.

Sounds good.

**Comment by mewmew** [comment]
# Some TODOs

## 1.

> I don't quite know myself what the following fields in Item mean:

> * drop_anim_update
> * inactive
> * drop_state and its enum
> * draw_quest_item

> And I do believe that some of them require comment at least. Actually I'm fine with leaving them as it is and later perhaps me or someone else will add comment or rename them if will find it appropriate.

I agree, lets leave them for now. We can always refine them once we learn more about the game engine.

## 2.

> Perhaps some other fields require comment too. Some enums may benefit from comments also, like for special effects or effect types adding item which has them as a point of reference, if they are rare. But possibly it could be added in the future.

I'm fine with adding additional comments later on. Still think you did a great job with the comments, and you clarified quite a few things that I didn't know.

## 3.

> Possibly I should remove part in comment about Q26.6 from health, mana bonus because it looks too obscure.

Could you elaborate on what Q26.6 fixed-point format is? Is it an established standard? I tried a quick Google and failed to come up with something. If it is an established standard like IEEE 754 floats, lets keep the comment. I'll read up about it.

## 4.

> I've tried adding \\ unused to effects missing in game but probably something longer like \\ missing in game should be used instead for more clarity.

We've also been using `unused` for now to indicate that no references in the game were found for it. However, going forward, we should probably using something along the lines of `// unused in win 1.09b, used in mac 1.09 and PSX jp`.

## 5.

> Should we really call reserved fields which are most likely padding? Should perhaps bit-fields be used for adding 24 - bit padding. (like int reserved : 24)?

I'm fine with removing these as I agree, they are most likely padding.

Perhaps we should add offset comments for members? That way it is clear that there may be space in between members.

## 6.

> Some enums clearly has underlying type that just 8-bit but some look like 32-bit (or were they just stored in int). It's important to denote that somehow, otherwise struct layout will be incomplete

Perhaps we can add something along the lines of

```c
   // offset: 014C (1 byte)
   item_effect_type prefix_effect_type;
   // offset: 014D (1 byte)
   item_effect_type suffix_effect_type;
```

No structure is larger than four bytes in hex, so offsets will always be within `0000` - `FFFF`.

## 7.

> Should I also used types like bool32_t for types which denote boolean values? I think that adding prefix is_ sort of makes clear that they represent bool.

We actually use this in our IDB database. It definitely makes sense to add that to the notes repo too.

File: enumerates.md [hunk 1]
```diff
@@ -83,6 +83,176 @@ typedef enum {
...
+    ITEM_EFFECT_TYPE_NO_STRENGTH_REQUIREMENT = 65,
+    ITEM_EFFECT_TYPE_SET_SPELL_ID_AND_CHARGES = 66, // (most likely erroneously) sets current spell charges to spell_id
+    ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED_3 = 67, // unused, same as ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED with value 3
+    ITEM_EFFECT_TYPE_MAKE_WEAPON_ONE_HANDED = 68,
+    ITEM_EFFECT_TYPE_3X_DAMAGE_VS_DEMONS = 69,
+    ITEM_EFFECT_TYPE_ALL_RESISTANCES_EQUAL_0 = 70,
+    ITEM_EFFECT_TYPE_CONSTANTLY_LOSE_HIT_POINTS = 72,
+    ITEM_EFFECT_TYPE_LIFE_STEAL_RANDOM = 73, // in range 0-12.5%, used for The Undead Crown
+    ITEM_EFFECT_TYPE_CONSTANT_INFRAVISION = 74, // unused
+    ITEM_EFFECT_TYPE_CUSTOM_ARMOR_CLASS = 75, // unused
```
Comment (mewmew) [inline enumerates.md]: Good catch! This was incorrectly named POSITIVE_ARMOR_CLASS, when CUSTOM_ARMOR_CLASS is a correct description.

Comment (mewmew) [inline enumerates.md]: This field is actually used. Remove the `// unused` comment.

From `items_set_effect` (addr: `0x420B68`).
```c
    case 76:
      items[v8].HP = (players[player_own_num].field_543C
                    + players[player_own_num].field_5430
                    + players[player_own_num].cur_dex / 5) << 6;
      break;
    case 77:
      items[v8].armor_class += (players[player_own_num].mp_base_max >> 6) / 10;
      break;
    case 78:
      v41 = 30 - players[player_own_num].clvl;
      v42 = &items[v8].fire_resistance;
      *v42 = v41;
      if ( v41 < 0 )
        *v42 = 0;
      break;
    case 79:
      v12 = &items[v8].armor_class;
```

For the same reason, remove the `// unused` comments from 76, 77, and 78.

File: enumerates.md [hunk 2]
```diff
@@ -83,6 +83,176 @@ typedef enum {
...
+    ITEM_EFFECT_TYPE_CHANGE_DURABILITY = 64,
+    ITEM_EFFECT_TYPE_NO_STRENGTH_REQUIREMENT = 65,
+    ITEM_EFFECT_TYPE_SET_SPELL_ID_AND_CHARGES = 66, // (most likely erroneously) sets current spell charges to spell_id
+    ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED_3 = 67, // unused, same as ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED with value 3
+    ITEM_EFFECT_TYPE_MAKE_WEAPON_ONE_HANDED = 68,
+    ITEM_EFFECT_TYPE_3X_DAMAGE_VS_DEMONS = 69,
+    ITEM_EFFECT_TYPE_ALL_RESISTANCES_EQUAL_0 = 70,
+    ITEM_EFFECT_TYPE_CONSTANTLY_LOSE_HIT_POINTS = 72,
+    ITEM_EFFECT_TYPE_LIFE_STEAL_RANDOM = 73, // in range 0-12.5%, used for The Undead Crown
+    ITEM_EFFECT_TYPE_CONSTANT_INFRAVISION = 74, // unused
```
Comment (mewmew) [inline enumerates.md]: It is used in `0x420B68`,

```c
    case 74:
      items[v8].special_effects |= 1u;
      break;
```

File: enumerates.md [hunk 3]
```diff
@@ -83,6 +83,176 @@ typedef enum {
...
+    ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED = 58,
+    ITEM_EFFECT_TYPE_FASTER_HIT_RECOVERY = 59,
+    ITEM_EFFECT_TYPE_FAST_BLOCK = 60,
+    ITEM_EFFECT_TYPE_ATTACK_DAMAGE_PLUS = 61,
+    ITEM_EFFECT_TYPE_RANDOM_SPEED_ARROWS = 62,
+    ITEM_EFFECT_TYPE_MIN_AND_MAX_ATTACK_BONUS = 63, // used for Butcher's Cleaver
+    ITEM_EFFECT_TYPE_CHANGE_DURABILITY = 64,
+    ITEM_EFFECT_TYPE_NO_STRENGTH_REQUIREMENT = 65,
+    ITEM_EFFECT_TYPE_SET_SPELL_ID_AND_CHARGES = 66, // (most likely erroneously) sets current spell charges to spell_id
+    ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED_3 = 67, // unused, same as ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED with value 3
```
Comment (mewmew) [inline enumerates.md]: It is used in `0x420B68`,

```c
    case 67:
      BYTE2(items[v8].special_effects) |= 8u;
      break;
```

> same as ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED with value 3

More specifically, it is the same as `ITEM_SPECIAL_EFFECT_FASTER_ATTACK_3`.

File: enumerates.md [hunk 4]
```diff
@@ -83,6 +83,176 @@ typedef enum {
...
+	ITEM_CODE_POTION_OF_FULL_REJUVENATION = 19,
+	ITEM_CODE_USE_LAST = 20, // unused
+	ITEM_CODE_SCROLL = 21,
+	ITEM_CODE_SCROLL_WITH_TARGET = 22,
+	ITEM_CODE_STAFF = 23,
+	ITEM_CODE_BOOK = 24,
+	ITEM_CODE_RING = 25,
+	ITEM_CODE_AMULET = 26,
+	ITEM_CODE_UNIQUE = 27,
+	ITEM_CODE_POTION_OF_HEALING_SOMETHING = 28, // unused
```
Comment (mewmew) [inline enumerates.md]: Update the comment to `// unused in win 1.09b, used in win 1.00 debug`.

From the debug release:

```c
    case ITEM_CODE_POTION_OF_HEALING_SOMETHING:
      nPlayer_num_W4 = nPlayer_num;
      pnHP_max = &gaPlayer[nPlayer_num].nHP_max;
      nHP_max_div_256 = gaPlayer[nPlayer_num].nHP_max >> 8;
      nHP_max_div_256_W = nHP_max_div_256;
      LOBYTE(nHP_max_div_256) = nHP_max_div_256 & 0xFE;// clear least significant bit.
      nHP_delta = (engine_rand_cap(39, nHP_max_div_256_W) << 6) + 32 * nHP_max_div_256;
      bPlayer_class_W = gaPlayer[nPlayer_num_W4].bPlayer_class;
      if ( bPlayer_class_W == PLAYER_CLASS_WARRIOR )// PLAYER_CLASS_WARRIOR
        nHP_delta *= 2;                         // bonus based on class. Warrior: 2 * nHP_delta
      if ( bPlayer_class_W == PLAYER_CLASS_ROGUE )
        nHP_delta += nHP_delta >> 1;            // bonus based on class. Rogue: 1.5 * nHP_delta
      nHP_cur = nHP_delta + gaPlayer[nPlayer_num_W4].nHP_cur;
      gaPlayer[nPlayer_num_W4].nHP_cur = nHP_cur;
      if ( *pnHP_max < nHP_cur )
        gaPlayer[nPlayer_num_W4].nHP_cur = *pnHP_max;
      nHP_base_cur = gaPlayer[nPlayer_num_W4].nHP_base_cur + nHP_delta;
      gaPlayer[nPlayer_num_W4].nHP_base_cur = nHP_base_cur;
      nHP_base_max = gaPlayer[nPlayer_num_W4].nHP_base_max;
      if ( nHP_base_cur > nHP_base_max )
        gaPlayer[nPlayer_num_W4].nHP_base_cur = nHP_base_max;
      gboolUpdate_health = TRUE;
      break;
```

Can you think of a clear name to capture the above code?

Comment (Predelnik) [inline enumerates.md]: Sadly I haven't got to corresponding code for normal potion yet but according to Jarulf's - normal potion is (1/8-3/8 * class_coeff) hp and this one looks like (1/8-2/8 * class_coeff) hp if I'm not missing anything
If it's missing from game, maybe just `POTION_OF_HEALING_OLD`

File: enumerates.md [hunk 5]
```diff
@@ -83,6 +83,176 @@ typedef enum {
...
+    ITEM_EFFECT_TYPE_DAMAGES_TARGETS_ARMOR = 57,
+    ITEM_EFFECT_TYPE_FASTER_ATTACK_SPEED = 58,
+    ITEM_EFFECT_TYPE_FASTER_HIT_RECOVERY = 59,
+    ITEM_EFFECT_TYPE_FAST_BLOCK = 60,
+    ITEM_EFFECT_TYPE_ATTACK_DAMAGE_PLUS = 61,
+    ITEM_EFFECT_TYPE_RANDOM_SPEED_ARROWS = 62,
+    ITEM_EFFECT_TYPE_MIN_AND_MAX_ATTACK_BONUS = 63, // used for Butcher's Cleaver
+    ITEM_EFFECT_TYPE_CHANGE_DURABILITY = 64,
+    ITEM_EFFECT_TYPE_NO_STRENGTH_REQUIREMENT = 65,
+    ITEM_EFFECT_TYPE_SET_SPELL_ID_AND_CHARGES = 66, // (most likely erroneously) sets current spell charges to spell_id
```
Comment (Predelnik) [inline enumerates.md]: Yeah but Jarulf also mentions that at one place. So when I got there I was like yeah that's a known bug :)

File: enumerates.md (hunk 3 above)
Comment (Predelnik) [inline enumerates.md]: Yeah special effect could be mentioned but I wanted to convey that the same can be achieved with other effect type with specific value so it's sort of redundant effect type. I'm fine with renaming it to swing I was just really keen on pointing out it's pointlessness :)

File: enumerates.md (hunk 2 above)
Comment (Predelnik) [inline enumerates.md]: Yeah what I meant by unused (in this and other cases) is that it's missing on any game items from and could surprise a person who knows the game but not its internals. That's why I think maybe these 2 cases should be different.

File: enumerates.md [hunk 6]
```diff
@@ -262,6 +550,53 @@ typedef enum {
...
+References:
+* https://github.com/sanctuary/notes/blob/master/data/structures.md#Item
+
+## item_special_effect
+
+```c
+// Bit flags used to describe unique, usually non-quantifiable effect on a player caused by items.
+typedef enum {
+    ITEM_SPECIAL_EFFECT_NONE = 0,
+    ITEM_SPECIAL_EFFECT_CONSTANT_INFRAVISION = 1, // unused
```
Comment (Predelnik) [inline enumerates.md]: Okay. I just wanted to point out difference to normal infravision.

File: structures.md [hunk 7]
```diff
@@ -24,6 +24,102 @@ typedef struct {
 } GossipData;
 ```
 
+## Item
+
+```c
+// Item describes in-game state of any game item.
+typedef struct
+{
+    uint random_seed;
```
Comment (Predelnik) [inline structures.md]: In gardening :) I don't know for me it makes a purpose a bit clearer but I'm fine with removing.

File: structures.md [hunk 8]
```diff
@@ -24,6 +24,102 @@ typedef struct {
...
+    uint16_t unknown;
```
Comment (Predelnik) [inline structures.md]: Yeah it looks like it has to do with item generation source but I'm not yet ready to name it anyhow. CF - creation flags, perhaps.

File: structures.md [hunk 9]
```diff
@@ -24,6 +24,102 @@ typedef struct {
...
+    uint16_t reserved_0;
+    item_type type;
+    int row;
```
Comment (Predelnik) [inline structures.md]: Oh yeah personally I named them tile_x, tile_y which I used in freeablo but then I saw that col, row is in the notes already in many places so I renamed them without thinking much which is which.

File: structures.md [hunk 10]
```diff
@@ -24,6 +24,102 @@ typedef struct {
...
+    int max_attack_damage;
+    int armor_class;
+    item_special_effect special_effect_flags; // bitmask
+    item_code code;
+    spell_id staff_spell_id;
+    int cur_staff_charges;
+    int max_staff_charges;
+    int cur_durability;
+    int max_durability; // if it equals 255 then item is indestructible
+    int percent_dmg_done_bonus;
```
Comment (Predelnik) [inline structures.md]: Yeah that was a thing I mistakenly didn't rename. I thought like "damage done" is perfectly valid term but then I googled it, and it looks like I've got deluded somehow.

File: structures.md [hunk 11]
```diff
@@ -24,6 +24,102 @@ typedef struct {
...
+    int light_radius_bonus; // measured in 10% chunks
+    uint8_t spell_level_bonus;
+    uint8_t held_in_hand;
+    int16_t reserved_4;
+    unique_item_id unique_id;
+    int fire_damage_min;
+    int fire_damage_max;
+    int lightning_damage_min;
+    int lightning_damage_max;
+    int armor_penetration;
```
Comment (Predelnik) [inline structures.md]: So it comes from piercing/puncturing/bashing suffix and technically lowers target AC but ends up acting similar to "to hit" eventually. In hellfire they do something different according to Jarulf's

**Comment by Predelnik** [comment]
> I do not wish to use _t suffixes or m_ prefixes. Feels to C++ influenced.

Well I don't have many experience with C libraries, but libgit2 uses `_t` for enums but I'm not a huge of these things either.

> Could you elaborate on what Q26.6 fixed-point format is? Is it an established standard?
https://en.wikipedia.org/wiki/Q_(number_format)

So actually I just found it recently on wiki, basically it's just a fixed point format where 6 bit is used for fraction part (since 2^6 = 64) and the rest is used as integer part (26=32-6)

> Regarding fast, faster, and numeric suffixes.

Yeah looks like my judgment was affected by the fact that in code they just same effect type with values 0,1,2 etc. which gets converted to bit mask. Seemed weird to me at that moment.

File: enumerates.md (hunk 4 above)
Comment (Predelnik) [inline enumerates.md]: Nevermind, I miscalculated, actually it seems like the same formula. Then I guess it'd be interesting to compare it against code for normal healing potion.

**Comment by mewmew** [comment]
Hej @Predelnik!

Thanks a lot for this pull request. We are both very impressed with the quality of your information gathering and reversing!

Will be fun to continue this adventure together. Lets unlock its mysteries.

Cheers /u & i

**Comment by mewmew** [comment]
@Predelnik Just for public records, would you be willing to release your contribution into the public domain? The intention is to keep the [project source code in the public domain](https://github.com/sanctuary/notes#public-domain) to simplify license handling for the end user.

**Comment by Predelnik** [comment]
Sure! I thought that anyone making pull-reqiest implicitly agrees with project license.

**Comment by mewmew** [comment]
> Sure! I thought that anyone making pull-reqiest implicitly agrees with project license.

Thanks. Yes, I would hope so to, but better to be explicit when possible :)

**Pull request merged**